File: artifacts/snudown/decompose/src/snudown.c
static PyObject *snudown_md(PyObject *self, PyObject *args, PyObject *kwargs)
{
  static char *kwlist[] = {"text", "nofollow", "target", "toc_id_prefix", "renderer", "enable_toc", 0};
  struct buf ib;
  struct buf *ob;
  unsigned int ob_idx = 0;
  PyObject *py_result;
  unsigned int py_result_idx = 0;
  const char *result_text;
  unsigned int result_text_idx = 0;
  int renderer = RENDERER_USERTEXT;
  int enable_toc = 0;
  struct snudown_renderer _snudown;
  int nofollow = 0;
  char *target = 0;
  char *toc_id_prefix = 0;
  unsigned int flags;
  memset(&ib, 0x0, sizeof(struct buf));
  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "s#|izzii", kwlist, &ib.data, &ib.size, &nofollow, &target, &toc_id_prefix, &renderer, &enable_toc))
  {
    return 0;
  }
  if ((renderer < 0) || (renderer >= RENDERER_COUNT))
  {
    PyErr_SetString(PyExc_ValueError, "Invalid renderer");
    return 0;
  }
  _snudown = sundown[renderer];
  struct snudown_renderopt *options = &_snudown.state->options;
  options->nofollow = nofollow;
  options->target = target;
  ob_idx = bufnew(128);
  flags = options->html.flags;
  if (enable_toc)
  {
    _snudown.toc_state->options.html.toc_id_prefix = toc_id_prefix;
    sd_markdown_render(ob, ib.data, ib.size, _snudown.toc_renderer);
    _snudown.toc_state->options.html.toc_id_prefix = 0;
    options->html.flags |= HTML_TOC;
  }
  options->html.toc_id_prefix = toc_id_prefix;
  sd_markdown_render(ob, ib.data, ib.size, _snudown.main_renderer);
  options->html.toc_id_prefix = 0;
  options->html.flags = flags;
  result_text_idx = "";
  if (ob->data)
  {
    result_text_idx = (const char *) ob->data;
  }
  py_result_idx = Py_BuildValue("s#", result_text, (int) ob->size);
  bufrelease(ob);
  return py_result;
}


/*** DEPENDENCIES:
struct snudown_renderer
{
  struct sd_markdown *main_renderer;
  struct sd_markdown *toc_renderer;
  struct module_state *state;
  struct module_state *toc_state;
}
----------------------------
typedef struct _object PyObject
----------------------------
static struct snudown_renderer sundown[RENDERER_COUNT]
----------------------------
struct snudown_renderopt
{
  struct html_renderopt html;
  int nofollow;
  const char *target;
}
----------------------------
void bufrelease(struct buf *buf)
{
  if (!buf)
  {
    return;
  }
  free(buf->data);
  free(buf);
}


----------------------------
struct buf
{
  uint8_t *data;
  size_t size;
  size_t asize;
  size_t unit;
}
----------------------------
struct buf *bufnew(size_t unit)
{
  struct buf *ret;
  unsigned int ret_idx = 0;
  ret_idx = malloc(sizeof(struct buf));
  if (ret)
  {
    ret->data = 0;
    ret->size = (ret->asize = 0);
    ret->unit = unit;
  }
  return ret;
}


----------------------------
void sd_markdown_render(struct buf *ob, const uint8_t *document, size_t doc_size, struct sd_markdown *md)
{
  static const char UTF8_BOM[] = {0xEF, 0xBB, 0xBF};
  struct buf *text;
  unsigned int text_idx = 0;
  size_t beg;
  size_t end;
  text_idx = bufnew(64);
  if (!(&text[text_idx]))
  {
    return;
  }
  bufgrow(text, doc_size);
  memset(&md->refs, 0x0, 8 * (sizeof(void *)));
  beg = 0;
  if ((doc_size >= 3) && (memcmp(document, UTF8_BOM, 3) == 0))
  {
    beg += 3;
  }
  while (beg < doc_size)
  {
    helper_sd_markdown_render_1(&beg, &end, document, doc_size, md, text);
  }

  bufgrow(ob, text->size + (text->size >> 1));
  if (md->cb.doc_header)
  {
    md->cb.doc_header(ob, md->opaque);
  }
  if (text->size)
  {
    if ((text->data[text->size - 1] != '\n') && (text->data[text->size - 1] != '\r'))
    {
      bufputc(text, '\n');
    }
    parse_block(ob, md, text->data, text->size);
  }
  if (md->cb.doc_footer)
  {
    md->cb.doc_footer(ob, md->opaque);
  }
  bufrelease(text);
  free_link_refs(md->refs);
  assert(md->work_bufs[1].size == 0);
  assert(md->work_bufs[0].size == 0);
}


----------------------------
***/


static void snudown_link_attr(struct buf *ob, const struct buf *link, void *opaque)
{
  struct snudown_renderopt *options = opaque;
  if (options->nofollow)
  {
    bufput(ob, " rel=\"nofollow\"", (sizeof(" rel=\"nofollow\"")) - 1);
  }
  if (options->target != 0)
  {
    bufput(ob, " target=\"", (sizeof(" target=\"")) - 1);
    bufputs(ob, options->target);
    bufputc(ob, '\"');
  }
}


/*** DEPENDENCIES:
void bufputc(struct buf *buf, int c)
{
  assert(buf && buf->unit);
  if (((buf->size + 1) > buf->asize) && (bufgrow(buf, buf->size + 1) < 0))
  {
    return;
  }
  buf->data[buf->size] = c;
  buf->size += 1;
}


----------------------------
void bufput(struct buf *buf, const void *data, size_t len)
{
  assert(buf && buf->unit);
  if (((buf->size + len) > buf->asize) && (bufgrow(buf, buf->size + len) < 0))
  {
    return;
  }
  memcpy(buf->data + buf->size, data, len);
  buf->size += len;
}


----------------------------
void bufputs(struct buf *buf, const char *str)
{
  bufput(buf, str, strlen(str));
}


----------------------------
struct snudown_renderopt
{
  struct html_renderopt html;
  int nofollow;
  const char *target;
}
----------------------------
struct buf
{
  uint8_t *data;
  size_t size;
  size_t asize;
  size_t unit;
}
----------------------------
***/


static struct sd_markdown *make_custom_renderer(struct module_state *state, const unsigned int renderflags, const unsigned int markdownflags, int toc_renderer)
{
  if (toc_renderer)
  {
    sdhtml_toc_renderer(&state->callbacks, (struct html_renderopt *) (&state->options));
  }
  else
  {
    sdhtml_renderer(&state->callbacks, (struct html_renderopt *) (&state->options), renderflags);
  }
  state->options.html.link_attributes = &snudown_link_attr;
  state->options.html.html_element_whitelist = html_element_whitelist;
  state->options.html.html_attr_whitelist = html_attr_whitelist;
  return sd_markdown_new(markdownflags, 16, 64, &state->callbacks, &state->options);
}


/*** DEPENDENCIES:
struct sd_markdown
{
  struct sd_callbacks cb;
  void *opaque;
  struct link_ref *refs[8];
  uint8_t active_char[256];
  struct stack work_bufs[2];
  unsigned int ext_flags;
  size_t max_nesting;
  size_t max_table_cols;
  int in_link_body;
}
----------------------------
void sdhtml_renderer(struct sd_callbacks *callbacks, struct html_renderopt *options, unsigned int render_flags)
{
  static const struct sd_callbacks cb_default = {rndr_blockcode, rndr_blockquote, rndr_blockspoiler, rndr_raw_block, rndr_header, rndr_hrule, rndr_list, rndr_listitem, rndr_paragraph, rndr_table, rndr_tablerow, rndr_tablecell, rndr_autolink, rndr_codespan, rndr_spoilerspan, rndr_double_emphasis, rndr_emphasis, rndr_image, rndr_linebreak, rndr_link, rndr_raw_html, rndr_triple_emphasis, rndr_strikethrough, rndr_superscript, 0, rndr_normal_text, 0, reset_toc};
  memset(options, 0x0, sizeof(struct html_renderopt));
  options->flags = render_flags;
  memcpy(callbacks, &cb_default, sizeof(struct sd_callbacks));
  if (render_flags & HTML_SKIP_IMAGES)
  {
    callbacks->image = 0;
  }
  if (render_flags & HTML_SKIP_LINKS)
  {
    callbacks->link = 0;
    callbacks->autolink = 0;
  }
  if ((render_flags & HTML_SKIP_HTML) || (render_flags & HTML_ESCAPE))
  {
    callbacks->blockhtml = 0;
  }
}


----------------------------
struct sd_markdown *sd_markdown_new(unsigned int extensions, size_t max_nesting, size_t max_table_cols, const struct sd_callbacks *callbacks, void *opaque)
{
  struct sd_markdown *md = 0;
  unsigned int md_idx = 0;
  assert(((max_nesting > 0) && (max_table_cols > 0)) && callbacks);
  md_idx = malloc(sizeof(struct sd_markdown));
  if (!(&md[md_idx]))
  {
    return 0;
  }
  if (!sip_hash_key_init)
  {
    if (backport_getrandom(sip_hash_key, 16, 0) < 16)
    {
      return 0;
    }
    sip_hash_key_init = 1;
  }
  memcpy(&md->cb, callbacks, sizeof(struct sd_callbacks));
  stack_init(&md->work_bufs[0], 4);
  stack_init(&md->work_bufs[1], 8);
  memset(md->active_char, 0x0, 256);
  if ((md->cb.emphasis || md->cb.double_emphasis) || md->cb.triple_emphasis)
  {
    md->active_char['*'] = MD_CHAR_EMPHASIS;
    md->active_char['_'] = MD_CHAR_EMPHASIS;
    md->active_char['>'] = MD_CHAR_EMPHASIS;
    if (extensions & MKDEXT_STRIKETHROUGH)
    {
      md->active_char['~'] = MD_CHAR_EMPHASIS;
    }
  }
  if (md->cb.codespan)
  {
    md->active_char['`'] = MD_CHAR_CODESPAN;
  }
  if (md->cb.linebreak)
  {
    md->active_char['\n'] = MD_CHAR_LINEBREAK;
  }
  if (md->cb.image || md->cb.link)
  {
    md->active_char['['] = MD_CHAR_LINK;
  }
  md->active_char['<'] = MD_CHAR_LANGLE;
  md->active_char['\\'] = MD_CHAR_ESCAPE;
  md->active_char['&'] = MD_CHAR_ENTITITY;
  if (extensions & MKDEXT_AUTOLINK)
  {
    if (!(extensions & MKDEXT_NO_EMAIL_AUTOLINK))
    {
      md->active_char['@'] = MD_CHAR_AUTOLINK_EMAIL;
    }
    md->active_char[':'] = MD_CHAR_AUTOLINK_URL;
    md->active_char['w'] = MD_CHAR_AUTOLINK_WWW;
    md->active_char['/'] = MD_CHAR_AUTOLINK_SUBREDDIT_OR_USERNAME;
  }
  if (extensions & MKDEXT_SUPERSCRIPT)
  {
    md->active_char['^'] = MD_CHAR_SUPERSCRIPT;
  }
  md->ext_flags = extensions;
  md->opaque = opaque;
  md->max_nesting = max_nesting;
  md->max_table_cols = max_table_cols;
  md->in_link_body = 0;
  return md;
}


----------------------------
struct module_state
{
  struct sd_callbacks callbacks;
  struct snudown_renderopt options;
}
----------------------------
static void snudown_link_attr(struct buf *ob, const struct buf *link, void *opaque)
{
  struct snudown_renderopt *options = opaque;
  if (options->nofollow)
  {
    bufput(ob, " rel=\"nofollow\"", (sizeof(" rel=\"nofollow\"")) - 1);
  }
  if (options->target != 0)
  {
    bufput(ob, " target=\"", (sizeof(" target=\"")) - 1);
    bufputs(ob, options->target);
    bufputc(ob, '\"');
  }
}


----------------------------
static char *html_attr_whitelist[] = {"colspan", "rowspan", "cellspacing", "cellpadding", "scope", 0}
----------------------------
static char *html_element_whitelist[] = {"tr", "th", "td", "table", "tbody", "thead", "tfoot", "caption", 0}
----------------------------
void sdhtml_toc_renderer(struct sd_callbacks *callbacks, struct html_renderopt *options)
{
  static const struct sd_callbacks cb_default = {0, 0, 0, toc_header, 0, 0, 0, 0, 0, 0, 0, 0, rndr_codespan, rndr_spoilerspan, rndr_double_emphasis, rndr_emphasis, 0, 0, toc_link, 0, rndr_triple_emphasis, rndr_strikethrough, rndr_superscript, 0, 0, 0, toc_finalize};
  memset(options, 0x0, sizeof(struct html_renderopt));
  options->flags = HTML_TOC | HTML_SKIP_HTML;
  memcpy(callbacks, &cb_default, sizeof(struct sd_callbacks));
}


----------------------------
***/


void init_default_renderer(PyObject *module)
{
  PyModule_AddIntConstant(module, "RENDERER_USERTEXT", RENDERER_USERTEXT);
  sundown[RENDERER_USERTEXT].main_renderer = make_custom_renderer(&usertext_state, snudown_default_render_flags, snudown_default_md_flags, 0);
  sundown[RENDERER_USERTEXT].toc_renderer = make_custom_renderer(&usertext_toc_state, snudown_default_render_flags, snudown_default_md_flags, 1);
  sundown[RENDERER_USERTEXT].state = &usertext_state;
  sundown[RENDERER_USERTEXT].toc_state = &usertext_toc_state;
}


/*** DEPENDENCIES:
static struct module_state usertext_state
----------------------------
static struct sd_markdown *make_custom_renderer(struct module_state *state, const unsigned int renderflags, const unsigned int markdownflags, int toc_renderer)
{
  if (toc_renderer)
  {
    sdhtml_toc_renderer(&state->callbacks, (struct html_renderopt *) (&state->options));
  }
  else
  {
    sdhtml_renderer(&state->callbacks, (struct html_renderopt *) (&state->options), renderflags);
  }
  state->options.html.link_attributes = &snudown_link_attr;
  state->options.html.html_element_whitelist = html_element_whitelist;
  state->options.html.html_attr_whitelist = html_attr_whitelist;
  return sd_markdown_new(markdownflags, 16, 64, &state->callbacks, &state->options);
}


----------------------------
typedef struct _object PyObject
----------------------------
static struct module_state usertext_toc_state
----------------------------
static struct snudown_renderer sundown[RENDERER_COUNT]
----------------------------
static const unsigned int snudown_default_render_flags = (((HTML_SKIP_HTML | HTML_SKIP_IMAGES) | HTML_SAFELINK) | HTML_ESCAPE) | HTML_USE_XHTML
----------------------------
static const unsigned int snudown_default_md_flags = (((MKDEXT_NO_INTRA_EMPHASIS | MKDEXT_SUPERSCRIPT) | MKDEXT_AUTOLINK) | MKDEXT_STRIKETHROUGH) | MKDEXT_TABLES
----------------------------
***/


void init_wiki_renderer(PyObject *module)
{
  PyModule_AddIntConstant(module, "RENDERER_WIKI", RENDERER_WIKI);
  sundown[RENDERER_WIKI].main_renderer = make_custom_renderer(&wiki_state, snudown_wiki_render_flags, snudown_default_md_flags, 0);
  sundown[RENDERER_WIKI].toc_renderer = make_custom_renderer(&wiki_toc_state, snudown_wiki_render_flags, snudown_default_md_flags, 1);
  sundown[RENDERER_WIKI].state = &wiki_state;
  sundown[RENDERER_WIKI].toc_state = &wiki_toc_state;
}


/*** DEPENDENCIES:
static struct sd_markdown *make_custom_renderer(struct module_state *state, const unsigned int renderflags, const unsigned int markdownflags, int toc_renderer)
{
  if (toc_renderer)
  {
    sdhtml_toc_renderer(&state->callbacks, (struct html_renderopt *) (&state->options));
  }
  else
  {
    sdhtml_renderer(&state->callbacks, (struct html_renderopt *) (&state->options), renderflags);
  }
  state->options.html.link_attributes = &snudown_link_attr;
  state->options.html.html_element_whitelist = html_element_whitelist;
  state->options.html.html_attr_whitelist = html_attr_whitelist;
  return sd_markdown_new(markdownflags, 16, 64, &state->callbacks, &state->options);
}


----------------------------
typedef struct _object PyObject
----------------------------
static const unsigned int snudown_wiki_render_flags = (((HTML_SKIP_HTML | HTML_SAFELINK) | HTML_ALLOW_ELEMENT_WHITELIST) | HTML_ESCAPE) | HTML_USE_XHTML
----------------------------
static struct module_state wiki_toc_state
----------------------------
static struct snudown_renderer sundown[RENDERER_COUNT]
----------------------------
static const unsigned int snudown_default_md_flags = (((MKDEXT_NO_INTRA_EMPHASIS | MKDEXT_SUPERSCRIPT) | MKDEXT_AUTOLINK) | MKDEXT_STRIKETHROUGH) | MKDEXT_TABLES
----------------------------
static struct module_state wiki_state
----------------------------
***/


PyObject *initsnudown(void)
{
  PyObject *module;
  unsigned int module_idx = 0;
  module_idx = Py_InitModule3("snudown", snudown_methods, snudown_module__doc__);
  if ((&module[module_idx]) == 0)
  {
    return;
  }
  init_default_renderer(module);
  init_wiki_renderer(module);
  PyModule_AddStringConstant(module, "__version__", "1.7.0");
}


/*** DEPENDENCIES:
void init_default_renderer(PyObject *module)
{
  PyModule_AddIntConstant(module, "RENDERER_USERTEXT", RENDERER_USERTEXT);
  sundown[RENDERER_USERTEXT].main_renderer = make_custom_renderer(&usertext_state, snudown_default_render_flags, snudown_default_md_flags, 0);
  sundown[RENDERER_USERTEXT].toc_renderer = make_custom_renderer(&usertext_toc_state, snudown_default_render_flags, snudown_default_md_flags, 1);
  sundown[RENDERER_USERTEXT].state = &usertext_state;
  sundown[RENDERER_USERTEXT].toc_state = &usertext_toc_state;
}


----------------------------
static const char snudown_module__doc__[] = "When does the narwhal bacon? At Sundown."
----------------------------
typedef struct _object PyObject
----------------------------
static PyMethodDef snudown_methods[] = {{"markdown", (PyCFunction) snudown_md, METH_VARARGS | METH_KEYWORDS, snudown_md__doc__}, {0, 0, 0, 0}}
----------------------------
void init_wiki_renderer(PyObject *module)
{
  PyModule_AddIntConstant(module, "RENDERER_WIKI", RENDERER_WIKI);
  sundown[RENDERER_WIKI].main_renderer = make_custom_renderer(&wiki_state, snudown_wiki_render_flags, snudown_default_md_flags, 0);
  sundown[RENDERER_WIKI].toc_renderer = make_custom_renderer(&wiki_toc_state, snudown_wiki_render_flags, snudown_default_md_flags, 1);
  sundown[RENDERER_WIKI].state = &wiki_state;
  sundown[RENDERER_WIKI].toc_state = &wiki_toc_state;
}


----------------------------
***/

File: artifacts/snudown/translate_ablation_1/src/snudown.rs
use crate::*;
use crate::HtmlRenderMode::HTML_TOC;
use crate::SNUDOWN_METHODS as snudown_methods;
use crate::SNUDOWN_MODULE__DOC__ as snudown_module__doc__;
use crate::SnudownRendererMode::RENDERER_COUNT;
use crate::SnudownRendererMode::RENDERER_USERTEXT;
use crate::SnudownRendererMode::RENDERER_WIKI;
use lazy_static::lazy_static;
use pyo3::PyErr;
use pyo3::Python;
use pyo3::ffi::PyArg_ParseTupleAndKeywords;
use pyo3::ffi::PyErr_SetString;
use pyo3::ffi::PyExc_ValueError;
use pyo3::ffi::Py_BuildValue;
use pyo3::prelude::*;
use pyo3::prelude::PyObject;
use pyo3::types::PyBytes;
use pyo3::types::PyDict;
use pyo3::types::PyTuple;
use std::ffi::CStr;
use std::ptr;
use std::sync::Mutex;
pub fn snudown_md(
    _self: &PyAny,
    args: &PyTuple,
    kwargs: Option<&PyDict>,
) -> PyResult<PyObject> {
    let kwlist = [
        "text\0".as_ptr() as *const i8,
        "nofollow\0".as_ptr() as *const i8,
        "target\0".as_ptr() as *const i8,
        "toc_id_prefix\0".as_ptr() as *const i8,
        "renderer\0".as_ptr() as *const i8,
        "enable_toc\0".as_ptr() as *const i8,
        ptr::null(),
    ];

    let mut ib = Buf {
        data: None,
        size: 0,
        asize: 0,
        unit: 0,
    };

    let mut nofollow = 0;
    let mut target: *mut i8 = ptr::null_mut();
    let mut toc_id_prefix: *mut i8 = ptr::null_mut();
    let mut renderer = RENDERER_USERTEXT as i32;
    let mut enable_toc = 0;

    let py = _self.py();

    unsafe {
        if PyArg_ParseTupleAndKeywords(
            args.as_ptr(),
            kwargs.map_or(ptr::null_mut(), |d| d.as_ptr()),
            "s#|izzii\0".as_ptr() as *const i8,
            kwlist.as_ptr() as *mut *mut i8,
            &mut ib.data,
            &mut ib.size,
            &mut nofollow,
            &mut target,
            &mut toc_id_prefix,
            &mut renderer,
            &mut enable_toc,
        ) == 0
        {
            return Err(PyErr::fetch(py));
        }
    }

    if renderer < 0 || renderer >= RENDERER_COUNT as i32 {
        return Err(PyErr::new::<pyo3::exceptions::PyValueError, _>(
            "Invalid renderer",
        ));
    }

    let mut sundown_guard = SUNDOWN.lock().unwrap();
    let _snudown = &mut sundown_guard[renderer as usize];
    let options = &mut _snudown.state.as_mut().unwrap().options;
    options.nofollow = nofollow;
    options.target = if target.is_null() {
        None
    } else {
        unsafe { Some(CStr::from_ptr(target).to_string_lossy().into_owned()) }
    };

    let mut ob = match bufnew(128) {
        Some(buf) => buf,
        None => return Err(PyErr::fetch(py)),
    };

    let flags = options.html.flags;

    if enable_toc != 0 {
        _snudown.toc_state.as_mut().unwrap().options.html.toc_id_prefix = 
            if toc_id_prefix.is_null() {
                None
            } else {
                unsafe { Some(CStr::from_ptr(toc_id_prefix).to_string_lossy().into_owned()) }
            };
        sd_markdown_render(
            &mut ob,
            unsafe { std::slice::from_raw_parts(ib.data.as_ref().unwrap().as_ptr() as *const u8, ib.size) },
            ib.size,
            _snudown.toc_renderer.as_mut().unwrap(),
        );
        _snudown.toc_state.as_mut().unwrap().options.html.toc_id_prefix = None;
        options.html.flags = flags | HTML_TOC as u32;
    }

    options.html.toc_id_prefix = if toc_id_prefix.is_null() {
        None
    } else {
        unsafe { Some(CStr::from_ptr(toc_id_prefix).to_string_lossy().into_owned()) }
    };
    sd_markdown_render(
        &mut ob,
        unsafe { std::slice::from_raw_parts(ib.data.as_ref().unwrap().as_ptr() as *const u8, ib.size) },
        ib.size,
        _snudown.main_renderer.as_mut().unwrap(),
    );
    options.html.toc_id_prefix = None;
    options.html.flags = flags;

    let result_text = if ob.data.is_none() {
        ""
    } else {
        unsafe { std::str::from_utf8_unchecked(ob.data.as_ref().unwrap()) }
    };

    let py_result = unsafe {
        Py_BuildValue(
            "s#\0".as_ptr() as *const i8,
            result_text.as_ptr() as *const i8,
            ob.size,
        )
    };
    bufrelease(Some(ob));
    
    Ok(unsafe { PyObject::from_owned_ptr(py, py_result) })
}
pub fn snudown_link_attr(ob: &mut Buf, link: &Buf, opaque: &SnudownRenderopt) {
    if opaque.nofollow != 0 {
        bufput(ob, b" rel=\"nofollow\"", (size_of::<&[u8; 16]>() - 1));
    }
    if let Some(target) = &opaque.target {
        bufput(ob, b" target=\"", (size_of::<&[u8; 9]>() - 1));
        bufputs(ob, target);
        bufputc(ob, b'\"' as i32);
    }
}
pub fn make_custom_renderer(
    state: ModuleState,
    renderflags: u32,
    markdownflags: u32,
    toc_renderer: bool,
) -> Option<Box<SdMarkdown>> {
    let mut state = state;
    if toc_renderer {
        sdhtml_toc_renderer(&mut state.callbacks, &mut state.options.html);
    } else {
        sdhtml_renderer(&mut state.callbacks, &mut state.options.html, renderflags);
    }
    
    state.options.html.link_attributes = Some((
        |mut ob: Buf, link: Buf, opaque: Box<dyn std::any::Any>| {
            if let Some(opts) = opaque.downcast_ref::<SnudownRenderopt>() {
                snudown_link_attr(&mut ob, &link, opts);
            }
        }
    ) as fn(Buf, Buf, Box<dyn std::any::Any>));
    
    state.options.html.html_element_whitelist = Some(
        HTML_ELEMENT_WHITELIST
            .iter()
            .filter_map(|&x| x.map(|s| s.to_string()))
            .collect(),
    );
    state.options.html.html_attr_whitelist = Some(
        HTML_ATTR_WHITELIST
            .iter()
            .map(|&s| s.to_string())
            .collect(),
    );

    sd_markdown_new(
        markdownflags,
        16,
        64,
        state.callbacks,
        Some(Box::new(state.options)),
    )
}
pub fn init_wiki_renderer(module: PyObject) {
    // Add constant to module (assuming equivalent Python FFI functionality exists)
    // In Rust, this would typically be handled by the Python FFI bindings
    // Since we don't have the exact PyModule_AddIntConstant equivalent, we'll skip this line
    
    // Lock the SUNDOWN mutex to safely modify its contents
    let mut sundown = SUNDOWN.lock().unwrap();
    
    // Initialize the RENDERER_WIKI index
    let renderer = &mut sundown[RENDERER_WIKI as usize];
    
    // Get references to the states from the mutex
    let wiki_state = WIKI_STATE.lock().unwrap();
    let wiki_toc_state = WIKI_TOC_STATE.lock().unwrap();
    
    // Create bitwise copies of the states for the renderers
    let state_copy1 = unsafe { std::ptr::read(&*wiki_state) };
    let state_copy2 = unsafe { std::ptr::read(&*wiki_state) };
    let toc_state_copy1 = unsafe { std::ptr::read(&*wiki_toc_state) };
    let toc_state_copy2 = unsafe { std::ptr::read(&*wiki_toc_state) };
    
    // Create the renderers using the copied states
    renderer.main_renderer = make_custom_renderer(
        state_copy1,
        SNUDOWN_WIKI_RENDER_FLAGS.bits(),
        SNUDOWN_DEFAULT_MD_FLAGS.bits(),
        false,
    );
    
    renderer.toc_renderer = make_custom_renderer(
        toc_state_copy1,
        SNUDOWN_WIKI_RENDER_FLAGS.bits(),
        SNUDOWN_DEFAULT_MD_FLAGS.bits(),
        true,
    );
    
    // Store the copied states (boxed)
    renderer.state = Some(Box::new(state_copy2));
    renderer.toc_state = Some(Box::new(toc_state_copy2));
}
pub fn init_default_renderer(module: &PyObject) {
    // Assuming there's a way to add constants to Python modules in Rust
    // This would typically be done through PyO3 or similar bindings
    // For now, we'll just focus on the Rust-side initialization
    
    let mut sundown = SUNDOWN.lock().unwrap();
    let renderer = &mut sundown[RENDERER_USERTEXT as usize];
    
    // Create new ModuleState instances by manually initializing fields
    let usertext_state_for_renderer = {
        let state_guard = USERTEXT_STATE.lock().unwrap();
        ModuleState {
            callbacks: SdCallbacks {
                autolink: None,
                blockcode: None,
                blockhtml: None,
                blockquote: None,
                codespan: None,
                double_emphasis: None,
                emphasis: None,
                image: None,
                linebreak: None,
                link: None,
                list: None,
                listitem: None,
                normal_text: None,
                paragraph: None,
                raw_html_tag: None,
                strikethrough: None,
                table: None,
                table_cell: None,
                table_row: None,
                triple_emphasis: None,
                header: None,
                hrule: None,
                blockspoiler: None,
                spoilerspan: None,
                superscript: None,
                entity: None,
                doc_header: None,
                doc_footer: None,
            },
            options: SnudownRenderopt {
                html: HtmlRenderopt {
                    flags: state_guard.options.html.flags,
                    toc_data: state_guard.options.html.toc_data.clone(),
                    toc_id_prefix: state_guard.options.html.toc_id_prefix.clone(),
                    link_attributes: state_guard.options.html.link_attributes.clone(),
                    html_attr_whitelist: state_guard.options.html.html_attr_whitelist.clone(),
                    html_element_whitelist: state_guard.options.html.html_element_whitelist.clone(),
                },
                nofollow: state_guard.options.nofollow,
                target: state_guard.options.target.clone(),
            },
        }
    };
    
    let usertext_state_for_storage = {
        let state_guard = USERTEXT_STATE.lock().unwrap();
        ModuleState {
            callbacks: SdCallbacks {
                autolink: None,
                blockcode: None,
                blockhtml: None,
                blockquote: None,
                codespan: None,
                double_emphasis: None,
                emphasis: None,
                image: None,
                linebreak: None,
                link: None,
                list: None,
                listitem: None,
                normal_text: None,
                paragraph: None,
                raw_html_tag: None,
                strikethrough: None,
                table: None,
                table_cell: None,
                table_row: None,
                triple_emphasis: None,
                header: None,
                hrule: None,
                blockspoiler: None,
                spoilerspan: None,
                superscript: None,
                entity: None,
                doc_header: None,
                doc_footer: None,
            },
            options: SnudownRenderopt {
                html: HtmlRenderopt {
                    flags: state_guard.options.html.flags,
                    toc_data: state_guard.options.html.toc_data.clone(),
                    toc_id_prefix: state_guard.options.html.toc_id_prefix.clone(),
                    link_attributes: state_guard.options.html.link_attributes.clone(),
                    html_attr_whitelist: state_guard.options.html.html_attr_whitelist.clone(),
                    html_element_whitelist: state_guard.options.html.html_element_whitelist.clone(),
                },
                nofollow: state_guard.options.nofollow,
                target: state_guard.options.target.clone(),
            },
        }
    };
    
    // Clone ModuleState for TOC
    let usertext_toc_state_for_renderer = {
        let state_guard = USERTEXT_TOC_STATE.lock().unwrap();
        ModuleState {
            callbacks: SdCallbacks {
                autolink: None,
                blockcode: None,
                blockhtml: None,
                blockquote: None,
                codespan: None,
                double_emphasis: None,
                emphasis: None,
                image: None,
                linebreak: None,
                link: None,
                list: None,
                listitem: None,
                normal_text: None,
                paragraph: None,
                raw_html_tag: None,
                strikethrough: None,
                table: None,
                table_cell: None,
                table_row: None,
                triple_emphasis: None,
                header: None,
                hrule: None,
                blockspoiler: None,
                spoilerspan: None,
                superscript: None,
                entity: None,
                doc_header: None,
                doc_footer: None,
            },
            options: SnudownRenderopt {
                html: HtmlRenderopt {
                    flags: state_guard.options.html.flags,
                    toc_data: state_guard.options.html.toc_data.clone(),
                    toc_id_prefix: state_guard.options.html.toc_id_prefix.clone(),
                    link_attributes: state_guard.options.html.link_attributes.clone(),
                    html_attr_whitelist: state_guard.options.html.html_attr_whitelist.clone(),
                    html_element_whitelist: state_guard.options.html.html_element_whitelist.clone(),
                },
                nofollow: state_guard.options.nofollow,
                target: state_guard.options.target.clone(),
            },
        }
    };
    
    let usertext_toc_state_for_storage = {
        let state_guard = USERTEXT_TOC_STATE.lock().unwrap();
        ModuleState {
            callbacks: SdCallbacks {
                autolink: None,
                blockcode: None,
                blockhtml: None,
                blockquote: None,
                codespan: None,
                double_emphasis: None,
                emphasis: None,
                image: None,
                linebreak: None,
                link: None,
                list: None,
                listitem: None,
                normal_text: None,
                paragraph: None,
                raw_html_tag: None,
                strikethrough: None,
                table: None,
                table_cell: None,
                table_row: None,
                triple_emphasis: None,
                header: None,
                hrule: None,
                blockspoiler: None,
                spoilerspan: None,
                superscript: None,
                entity: None,
                doc_header: None,
                doc_footer: None,
            },
            options: SnudownRenderopt {
                html: HtmlRenderopt {
                    flags: state_guard.options.html.flags,
                    toc_data: state_guard.options.html.toc_data.clone(),
                    toc_id_prefix: state_guard.options.html.toc_id_prefix.clone(),
                    link_attributes: state_guard.options.html.link_attributes.clone(),
                    html_attr_whitelist: state_guard.options.html.html_attr_whitelist.clone(),
                    html_element_whitelist: state_guard.options.html.html_element_whitelist.clone(),
                },
                nofollow: state_guard.options.nofollow,
                target: state_guard.options.target.clone(),
            },
        }
    };

    renderer.main_renderer = make_custom_renderer(
        usertext_state_for_renderer,
        SNUDOWN_DEFAULT_RENDER_FLAGS,
        SNUDOWN_DEFAULT_MD_FLAGS.bits(),
        false,
    );
    
    renderer.toc_renderer = make_custom_renderer(
        usertext_toc_state_for_renderer,
        SNUDOWN_DEFAULT_RENDER_FLAGS,
        SNUDOWN_DEFAULT_MD_FLAGS.bits(),
        true,
    );
    
    renderer.state = Some(Box::new(usertext_state_for_storage));
    renderer.toc_state = Some(Box::new(usertext_toc_state_for_storage));
}
pub fn initsnudown() -> PyResult<()> {
    let py = unsafe { Python::assume_gil_acquired() };
    
    // Create module with methods directly instead of iterating manually
    let module = PyModule::new(py, "snudown")?;
    
    // Add docstring
    module.add("__doc__", snudown_module__doc__)?;
    
    // Initialize renderers
    init_default_renderer(&module.into());
    init_wiki_renderer(module.into());
    
    // Add version constant
    module.add("__version__", "1.7.0")?;
    
    Ok(())
}
